serialise/src/Codec/Serialise/GeneralisedUTF8.hs

```diff
@@ -0,0 +1,87 @@
+{-# LANGUAGE BangPatterns #-}
+
+module Codec.Serialise.GeneralisedUTF8 where
+
+import Control.Monad.ST
+import Data.Bits
+import Data.Char
+import Data.Word
+import Data.Primitive.ByteArray
+
+encodeGenUTF8 :: String -> ByteArray
+encodeGenUTF8 st = runST $ do
+    ba <- newByteArray (length st)
+    go ba 0 st
+  where
+    go :: MutableByteArray s -> Int -> [Char] -> ST s ByteArray
+    go ba !_    [] = unsafeFreezeByteArray ba
+    go ba !off  (c:cs)
+      | off + 4 >= cap = do
+        ba' <- newByteArray (cap + cap `div` 2 + 1)
+        copyMutableByteArray ba' 0 ba 0 off
```

Comment:
Notice how already we are going to need ByteArray with offset and length on the encoder side? You're growing the array here but we're not trimming it again at the end, and nor do we want to we just want to keep track of the final offset.